### ISSUE TITLE
randomized proving

### ISSUE THREAD

**Issue by strobelm** [CindyJS/CindyJS#296]
Thanks to @gagern the foundations of randomized proving are laid. Now we can think about how to implement them. 
1. save geo state: do we need a full save/restore system for that or is the current implementation sufficient? Perhaps we just want to copy csgeo.
2. stop drawing: do we want to disable drawing while we're testing the conjectures?
3. random walks: with the current implementation of tracing only small movements are possible so we just start to wobble around with the objects via movepointscr.
4. restore geo state

**Comment by kranich** [comment]
Please bear in mind that questions 1. and 2. also apply to locus generation.

**Comment by gagern** [comment]
1. theoretically we only need to copy [the state arrays](https://github.com/CindyJS/CindyJS/blob/0d762ae7d8858ba614ab363743f90c8aa2247d9b/src/js/libgeo/Tracing.js#L11-L20). If you restore that state and run `updatePosition` on all elements, then the previous state should be completely restored. There already is machinery in place to increase the number of state arrays, so we could add `"savedIn"` and `"savedGood"` to `stateArrayNames` and then store the backups in those. Currently this doesn't handle physics, but on the other hand we probably can completely ignore physics for the sake of theorem proving, and leave its state alone. See also [`backupGeo` and `restoreGeo`](https://github.com/CindyJS/CindyJS/blob/0d762ae7d8858ba614ab363743f90c8aa2247d9b/src/js/Setup.js#L307-L345) which do create backups including some physics information.
1. Definitely. The conjecture testing should be done synchroneously, so there should be no room for the drawing code to even get a chance at rendering anything.
1. I think geometric primitives should get a new method `randomParameter` which can be called to obtain a suitable parameter value as the target for the trace of a given free element.

I wonder whether we want something like

```
function withSavedState(savedIn, savedGood, toCall) {
    savedIn.set(stateIn);
    savedGood.set(stateLastGood);
    toCall.apply(this, Array.prototype.slice(arguments, 3)); // call with remaining args
    stateLastGood.set(savedGood);
    stateIn.set(savedIn);
}
```

so that we have all the machinery of backing up and restoring state in one place, and can use that for both randomized proving and locus creation.

**Comment by strobelm** [comment]
Hi i just hacked a first rough draft (ignoring the locus case currently, sorry @kranich ;-) ):
https://github.com/strobelm/CindyJS/tree/rand-prove

Looks like it we are not very far from our goal.

**Comment by gagern** [comment]
Do we want to wriggle one element at a time, or all simultaneously? In the former case, we have to make very sure that we don't stop in a singular situation when switching to a different mover. In the latter case, we can always start from the original position so degeneracy shouldn't be that much of a problem, but we'll need extra code to interpolate multiple elements at once. And for semi-free elements we won't be able to find coordinates which are valid in the target situation; I don't know how much of a problem that would be. In either case, I'd suggest we try to choose the new parameters *completely* at random, without referencing the current position. That way we become more independent of the current configuration and any numerical uncertainties it might involve.

**Comment by kortenkamp** [comment]
In Cinderella we wiggle them one after the other, not simultaneously.

It does not make sense to move away much (an epsilon environment is sufficient theoretically), and choosing parameters completely at random will be numerically difficult for the tracing algorithm.

**Comment by strobelm** [comment]
I also think wobbeling around the current position is the only practical
solution, we could run into really huge movements (not speaking about
points at l_\infty) if we choose them randomly.

As @gagern pointed out it's nevertheless important to care about
singularities. The probability of hitting one is of course 0 (almost
surely) but again we have to care about numerics. Furthermore snapping
should be disabled while we are run the prover.

Am 09.03.2016 um 10:00 schrieb kortenkamp:
>
> In Cinderella we wiggle them one after the other, not simultaneously.
>
> It does not make sense to move away much (an epsilon environment is
> sufficient theoretically), and choosing parameters completely at
> random will be numerically difficult for the tracing algorithm.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CindyJS/CindyJS/issues/296#issuecomment-194191462>.
>

-- 
Michael Strobel, M.Sc.

Technische Universität München -- Zentrum Mathematik
Lehrstuhl für Geometrie und Visualisierung (M10)
Boltzmannstr. 3, 85747 Garching bei München 
https://www-m10.ma.tum.de/bin/view/Lehrstuhl/MichaelStrobel

**Comment by gagern** [comment]
Snapping is (or should be) performed for mouse input only, and should not affect a programmatic change of parameters. We should not move pinned points, though, as they might have been pinned to special positions leading to additional and desired incidences.

I'm not convinced that large movements are indeed problematic. But if small changes work for Cinderella, they should work for us as well. Do we wiggle into the complex plane, or should we choose parameters to remain real if they initially are real? I'd guess that using complex detours should help us avoid some nasty special cases. On the other hand, things like “point on segment” only make sense with real parameters. Or should we disregard the restriction to the segment, and allow arbitrary positions on the line during theorem proving?

What do we do when we end up in a singularity? Keep moving the current element until we get out again? Or abort the move (perhaps after checking the conjectures) and move something else starting at the last good position? What if the initial situation is singular? Do we move at random, or do we try to deliberatly get out of the singular situation by moving some mover on which the singularity depends? How do we handle that in code? Can we ensure that we'll never loop forever trying to get out of the singular situation? For example, the construction might have constructed a tangent to a conic (using some suitable theorem and construction sequence), and intersecting the tangent with the conic will *always* be singular, no matter how long we wobble around. Should we treat statements like “this operation will *always* be singular” as conjectures in their own right? How many are there? Does that help us solve the problem of correctly dealing with singularities?

What if there are *many* movers (which affect a given conjecture)? If we only move each of them a bit, and only perform a fixed number of such moves I'm concerned that we don't randomize the result enough to actually make a statement, since we might not even have touched some of the movers.

**Comment by kortenkamp** [comment]
Snapping should never be use for programmatic changes, of course. Otherwise the whole concept of tracing doesn't make sense.

> I'm not convinced that large movements are indeed problematic.

They are, believe me.

> Do we wiggle into the complex plane[...]?

Yes, that increases the reliability (without proof).

> What do we do when we end up in a singularity?

We have to abort the move.

> What if the initial situation is singular?

We should not care, a first move will usually resolve that singularity. If not, then it's almost a proof that it is *always* singular.

> How do we handle that in code? 

In Cinderella we have a "don't care" flag that is initially set and unset in the case that there is no singularity. This works fine.

> What if there are many movers (which affect a given conjecture)?

In Cinderella we move all movers three times (i.e. if A, B, C are points affecting the conjecture, we move A, B, C, A, B, C, A, B, C, each by a small random step).

**Comment by strobelm** [comment]
I pushed quite a large update including more working examples (not well documented yet) if somebody is interested.

**Comment by strobelm** [comment]
Hi, now all operations which are marked as movable in geoOps are implemented. The only things which causes me headaches is that PointOnLine won't move if i call 

```
movepointscr(PoinOnLinePoint, newhomog, "homog");
```

Any hints on that?

**Comment by strobelm** [comment]
Never mind, i found the problem.

**Comment by strobelm** [comment]
Merged in #299.
